merlin/models/tf/outputs/contrastive.py

```diff
@@ -0,0 +1,345 @@
+)
+from merlin.models.tf.outputs.sampling.base import (
+    Candidate,
+    ItemSamplersType,
+    parse_negative_samplers,
+)
+from merlin.models.tf.typing import TabularData
+from merlin.models.tf.utils import tf_utils
+from merlin.models.utils import schema_utils
+from merlin.models.utils.constants import MIN_FLOAT
+from merlin.schema import ColumnSchema, Schema
+
+LOG = logging.getLogger("merlin_models")
+
+
+@tf.keras.utils.register_keras_serializable(package="merlin.models")
+class ContrastiveOutput(ModelOutput):
+    """Categorical output
+
+    Parameters
+    ----------
+    prediction: Union[Schema, ColumnSchema,
+                EmbeddingTable, 'CategoricalTarget',
+                'EmbeddingTablePrediction']
+        The target feature to predict. To perform weight-tying [1] technique, you should provide
+        the `EmbeddingTable` or `EmbeddingTablePrediction` related to the
+        target feature.
+    negative_samplers: ItemSamplersType
+        List of samplers for negative sampling,
+        by default None
+    pre: Optional[Block], optional
+        Optional block to transform predictions before computing the binary logits,
+        by default None
+    post: Optional[Block], optional
+        Optional block to transform the binary logits,
+        by default None
+    logits_temperature: float, optional
+        Parameter used to reduce model overconfidence, so that logits / T.
+        by default 1
+    name: str, optional
+        The name of the task, by default None
+    default_loss: Union[str, tf.keras.losses.Loss], optional
+        Default loss to use for categorical-classification
+        by default 'categorical_crossentropy'
+    get_default_metrics: Callable, optional
+        A function returning the list of default metrics
+        to use for categorical-classification
+
+    References:
+    ----------
+    [1] Hakan Inan, Khashayar Khosravi, and Richard Socher. 2016. Tying word vectors
+    and word classifiers: A loss framework for language modeling. arXiv preprint
+    arXiv:1611.01462 (2016).
+
+    """
+
+    def __init__(
+        self,
+        to_call: Union[
+            Schema,
+            ColumnSchema,
+            EmbeddingTable,
+            CategoricalTarget,
+            EmbeddingTablePrediction,
+            DotProduct,
+        ],
+        negative_samplers: ItemSamplersType,
+        target_name: str = None,
+        pre: Optional[Layer] = None,
+        post: Optional[Layer] = None,
+        logits_temperature: float = 1.0,
+        name: Optional[str] = None,
+        default_loss: Union[str, tf.keras.losses.Loss] = "categorical_crossentropy",
+        default_metrics_fn: MetricsFn = default_categorical_prediction_metrics,
+        downscore_false_negatives=True,
+        false_negative_score: float = MIN_FLOAT,
+        query_name: str = "query",
+        candidate_name: str = "candidate",
+        **kwargs,
+    ):
+        _to_call = kwargs.pop("to_call", None)
```

Comment:
I realize this is just existing code with the variables renamed, but I'm wondering if it is possible to have `to_call` as a keyword argument when the first argument is `to_call`?
```python
>>> def foo(to_call, **kwargs):
...     _to_call = kwargs.pop("to_call", None)
...     print(_to_call)
... 
>>> foo(None, to_call="bar")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() got multiple values for argument 'to_call'
```

Reply:
Good point